File: sources/opencv-samples/6_CameraCapture/main.cpp
// https://www.selfmadetechie.com/how-to-create-a-webcam-video-capture-using-opencv-c
// https://learnopencv.com/read-write-and-display-a-video-using-opencv-cpp-python/

#include <stdio.h>
#include <csignal>
#include <iostream>
#include <sys/stat.h>
#include <opencv2/opencv.hpp>
using namespace cv;
using namespace std;

bool hasToBreak = false;

void signal_handler(int signal)
{
    hasToBreak = true;
}

int main(int argc, char** argv)
{   
    // Install a signal handler
    std::signal(SIGINT, signal_handler);

    if ( argc > 2 )
    {
        printf("Usage: ./main.out [g|G]\n");
        printf("Read the README.md\n");
        return -1;
    }

    printf("Trying to open /dev/video0 with CAP_V4L2\n");
    VideoCapture cap(0, cv::CAP_V4L2);  // Add cv::CAP_V4L2 to fix: Embedded video playback halted; module v4l2src0 reported: Failed to allocate required memory.

    if (!cap.isOpened()) 
    {
        cout << "Can not read the device" << endl;
        return 0;
    }

    bool convertToGray = false;
    if (argc == 2)
    {
        if (!strcmp("g", argv[1]) || !strcmp("G", argv[1]))
        {
            convertToGray = true;
        }
        else
        {
            printf("Unexpected parameter given\n");
            printf("Read the README.md\n");
            return -1;
        }
    }

    std::time_t rawtime;
    std::tm* timeinfo;
    char buffer [80];

    std::time(&rawtime);
    timeinfo = std::localtime(&rawtime);

    std::strftime(buffer, 80, "%Y-%m-%d %H:%M:%S", timeinfo);

    string screenshot_name = buffer;
    uint screenshot_count = 0;

    string window_name = "Camera Capture";
    char k;
    namedWindow(window_name);
    waitKey(500);

    while (1)
    {
        Mat frame;
        cap >> frame;

        if (!frame.empty()) 
        {
            if (convertToGray) 
            {
                cvtColor(frame, frame, cv::COLOR_RGB2GRAY);
            }
            imshow(window_name, frame);
            k = (char)waitKey(1);
        }
        else
        {
            printf("frame is empty\n");
            k = (char)waitKey(500);
        }    
        
        // break the loop if
        if (hasToBreak ||                                                   // due to Ctrl+C
            k == 27 ||                                                      // ESC key
            cv::getWindowProperty(window_name, WND_PROP_AUTOSIZE) == -1)    // window is closed
        {
            break;
        }
        if (k == 83 || k == 115)   // S | s
        {
            if (!frame.empty())
            {
                screenshot_count++;
                string path = "./screenshot/" + screenshot_name + "_" + std::to_string(screenshot_count) + ".jpg";
                mkdir("./screenshot", 0777);
                imwrite(path, frame);
                cout << path << endl;
            }
        }
    }

    printf("Capture done\n");
    cap.release();
    cv::destroyAllWindows();
    return 0;
}

File: sources/opencv-samples/6_CameraCapture/main.py
# https://learnopencv.com/read-write-and-display-a-video-using-opencv-cpp-python/

import cv2 as cv
import argparse
import sys
import signal
import os
from datetime import datetime


global has_to_break


def sigint_handler(signal, frame):
    global has_to_break
    has_to_break = True


def main():
    global has_to_break
    has_to_break = False
    
    args = parser.parse_args()

    print("Trying to open /dev/video0 with CAP_V4L2")
    cap = cv.VideoCapture(0, cv.CAP_V4L2)  # Add cv::CAP_V4L2 to fix: Embedded video playback halted; module v4l2src0 reported: Failed to allocate required memory.

    if not cap.isOpened():
        sys.exit("Can not read the device")

    print("Press ESC to quit and S to take a screenshot")
    screenshot_count = 0
    screenshot_name = datetime.today().strftime('%Y-%m-%d %H:%M:%S')
    window_name = "Camera Capture"
    cv.namedWindow(window_name)
    cv.waitKey(500)
    k = 0
    while True:
          
        ret, frame = cap.read()
        
        if not (ret == False or frame is None):
            if args.grayscale:
                frame = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)
            cv.imshow(window_name, frame)
            k = cv.waitKey(1) & 0xFF

        else:
            print("frame is empty")
            k = cv.waitKey(500) & 0xFF

        # break the loop if Ctrl+C or key ESC
        # cv.getWindowProperty(window_name, cv.WND_PROP_VISIBLE) == -1:  # window is closed
        if has_to_break or k == 27:
            break

        if k == 83 or k == 115:   # S | s
            if frame is not None:
                screenshot_count += 1
                path = f"./screenshot/{screenshot_name}_{screenshot_count}.jpg"
                if not os.path.exists("./screenshot"):
                    os.mkdir("./screenshot")
                cv.imwrite(path, frame)
                print(path)

    print("Capture done")
    # After the loop release the cap object
    cap.release()

    # Destroy all the windows
    cv.destroyAllWindows()


if __name__ == '__main__':
    signal.signal(signal.SIGINT, sigint_handler)
    parser = argparse.ArgumentParser(description="Camera Capture main.py CLI")
    parser.add_argument("-g", "--grayscale", action="store_true", help="capture the camera with GRAYSCALE")
    main()
